File: 14/Solution.java
import java.util.Arrays;
import java.util.stream.IntStream;
import java.util.stream.Collectors;
import static java.lang.System.out;

class Solution {
    public static String longestCommonPrefix(String[] strs) {
        return Arrays.stream(strs).reduce(strs[0], (acc, s) -> {
            if (acc.startsWith(s)) return s;
            return IntStream.range(0, acc.length() < s.length() ? acc.length() : s.length())
                .takeWhile(i -> acc.charAt(i) == s.charAt(i))
                .mapToObj(i -> String.valueOf(acc.charAt(i)))
                .collect(Collectors.joining())
                .toString();
        });
    }

    public static void main(String[] args) {
        out.println(longestCommonPrefix(new String[]{"flower", "flow", "flight"}));
        out.println(longestCommonPrefix(new String[]{"Dog", "racecar", "car"}));
        out.println(longestCommonPrefix(new String[]{"ab", "a"}));
    }
}

File: 14/solution.rs
pub fn longest_common_prefix(strs: Vec<&str>) -> String {
    strs.iter()
        .map(|s| s.to_string())
        .reduce(|acc, s| {
            acc.chars().zip(s.chars())
                .take_while(|(c, k)| c == k)
                .map(|(c, _)| c)
                .collect()
        })
        .unwrap_or_else(|| String::from(""))
}

fn main() {
    println!("{}", longest_common_prefix(vec!["flower", "flow", "flight"]));
    println!("{}", longest_common_prefix(vec!["dog", "racecar", "car"]));
    println!("{}", longest_common_prefix(vec!["ab", "a"]));
}
